### PR TITLE
Unitful annotations satisfying compressible Navier Stokes tests

### PR DESCRIPTION
**Pull request by rohanmclure**
As it stands, this branch correctly implements the compressible navier stokes tests for CPU (while permitting non-unitful drivers to coexist), however the following issues remain outstanding:

- `PlanetParameters` will always provide unitful parameters. May need to completely duplicated methods to remove these unit annotations where not in use.
- Use of unitful / non-unitful `Union` types in structure definitions prohibits us from running `MoistThermodynamics` code on GPU's - it causes issues with type inference and so requires dynamic dispatch.

### CODE REVIEW

**Comment by tapios** [comment]
@charleskawczynski Could you please review this?

**Comment by charleskawczynski** [comment]
If possible, I’d like to vote replacing all string, e.g. u”kg*m/s^2”, with a method call/variable that will return the units string (‘unit_force=u”kg*m/s^2”’)

I have a couple reasons for this request. The first is that verifying the correctness of “kg*m/s^2” requires more work than verifying the correctness of “unit_force”. Second, if we ever needed to do a replace all (or just make all units unity), it’s far easier to do if all of these string definitions are done in one place (we may not catch all “kg*m” vs “m*kg”).

File: src/Common/MoistThermodynamics/MoistThermodynamics.jl [hunk 1]
```diff
@@ -464,22 +470,22 @@ relation to obtain the saturation vapor pressure `p_v_sat` as a function of
...
+saturation_vapor_pressure(T::U(FT,:temp), ::Ice) where {FT<:Real} =
   saturation_vapor_pressure(T, FT(LH_s0), FT(cp_v) - FT(cp_i))
 saturation_vapor_pressure(ts::ThermodynamicState{FT}, ::Ice) where {FT<:Real} =
   saturation_vapor_pressure(air_temperature(ts), FT(LH_s0), FT(cp_v) - FT(cp_i))
 
-function saturation_vapor_pressure(T::FT, LH_0::FT, Δcp::FT) where {FT<:Real}
+function saturation_vapor_pressure(T::U(FT,:temp), LH_0::U(FT,:latent), Δcp::U(FT,:shc)) where {FT<:Real}
 
     return FT(press_triple) * (T/FT(T_triple))^(Δcp/FT(R_v)) *
-        exp( (FT(LH_0) - Δcp*FT(T_0))/FT(R_v) * (1 / FT(T_triple) - 1 / T) )
```
Comment (charleskawczynski) [inline src/Common/MoistThermodynamics/MoistThermodynamics.jl]: Please add `FT` back in

File: src/Common/MoistThermodynamics/MoistThermodynamics.jl [hunk 2]
```diff
@@ -534,7 +540,7 @@ function q_vap_saturation(T::FT, ρ::FT, q::PhasePartition{FT}=q_pt_0(FT)) where
     Δcp     = _liquid_frac * (FT(cp_v) - FT(cp_l)) + _ice_frac * (FT(cp_v) - FT(cp_i))
 
     # saturation vapor pressure over possible mixture of liquid and ice
-    p_v_sat = saturation_vapor_pressure(T, FT(LH_0), Δcp)
+    p_v_sat = saturation_vapor_pressure(T, LH_0, Δcp)
```
Comment (charleskawczynski) [inline src/Common/MoistThermodynamics/MoistThermodynamics.jl]: Please add `FT` back in

**Comment by rohanmclure** [comment]
bors try

File: src/Arrays/MPIStateArrays.jl [hunk 3]
```diff
@@ -159,6 +159,7 @@ function MPIStateArray{FT}(mpicomm, DA, Np, nstate, numelem;
                           ) where {FT}
 
   if weights == nothing
+    @show DA
```
Comment (charleskawczynski) [inline src/Arrays/MPIStateArrays.jl]: Remove @show?

File: src/Atmos/Model/linear.jl [hunk 4]
```diff
@@ -12,9 +13,15 @@ and, optionally,
...
 """
-function linearized_air_pressure(ρ::FT, ρe_tot::FT, ρe_pot::FT,
-                                 ρq_tot::FT=FT(0), ρq_liq::FT=FT(0), ρq_ice::FT=FT(0)) where {FT<:Real}
-  ρ*FT(R_d)*FT(T_0) + FT(R_d)/FT(cv_d)*(ρe_tot - ρe_pot - (ρq_tot - ρq_liq)*FT(e_int_v0) + ρq_ice*(FT(e_int_i0) + FT(e_int_v0)))
+#FIXME
+function linearized_air_pressure(ρ::U(FT,:density), ρe_tot::U(FT,:energypv), ρe_pot::U(FT,:energypv),
+                                 ρq_tot::U(FT,:density)=FT(0)*get_unit(ρ isa Quantity, :density),
+                                 ρq_liq::U(FT,:density)=FT(0)*get_unit(ρ isa Quantity, :density),
+                                 ρq_ice::U(FT,:density)=FT(0)*get_unit(ρ isa Quantity, :density)) where {FT<:Real}
+  u = ρ isa Quantity #FIXME
```
Comment (charleskawczynski) [inline src/Atmos/Model/linear.jl]: Can you include a more descriptive #FIXME here?

File: src/Atmos/Model/moisture.jl [hunk 5]
```diff
@@ -66,10 +66,10 @@ Assumes the moisture components are computed via thermodynamic equilibrium.
 Base.@kwdef struct EquilMoist <: MoistureModel
   maxiter::Int = 3
 end
-vars_state(::EquilMoist,FT) = @vars(ρq_tot::FT)
-vars_gradient(::EquilMoist,FT) = @vars(q_tot::FT, h_tot::FT)
-vars_diffusive(::EquilMoist,FT) = @vars(∇q_tot::SVector{3,FT})
-vars_aux(::EquilMoist,FT) = @vars(temperature::FT, θ_v::FT, q_liq::FT)
+vars_state(m::EquilMoist,FT) = @uvars(m, ρq_tot::U(FT,:density))
+vars_gradient(m::EquilMoist,FT) = @uvars(m, q_tot::FT, h_tot::U(FT,:gravpot))
+vars_diffusive(m::EquilMoist,FT) = @uvars(m, ∇q_tot::SVector{3, U(FT,u"m^-1")})
```
Comment (charleskawczynski) [inline src/Atmos/Model/moisture.jl]:
```suggestion
vars_diffusive(m::EquilMoist,FT) = @uvars(m, ∇q_tot::SVector{3, U(FT,u:invmeters)})
```
?

**Comment by simonbyrne** [comment]
bors try

File: src/Atmos/Model/moisture.jl (hunk 5 above)
Comment (charleskawczynski) [reply src/Atmos/Model/moisture.jl]: Or can we somehow use `:meters`?

**Comment by rohanmclure** [comment]
bors try

**Comment by rohanmclure** [comment]
bors try

**Comment by rohanmclure** [comment]
bors try

**Comment by charleskawczynski** [comment]
I'm not sure we have the bandwidth to revive this at the moment. Also, this may require re-thinking if we break of moist thermodynamics as planned.

**Pull request closed without merging**